src/v/archival/segment_reupload.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "cloud_storage/types.h"
+#include "model/fundamental.h"
+
+namespace cloud_storage {
+class partition_manifest;
+} // namespace cloud_storage
+
+namespace storage {
+class disk_log_impl;
+class ntp_config;
+class segment;
+} // namespace storage
```

Comment:
Is this a well known pattern? I tried searching for it but couldn't find much. Could you point out some resources about it? I am kind of new to C++ so it would be good to learn more about it and use it here if possible.

Reply:
I think Vlad is referring to moving all these forward declarations to a fwd.h header like it's done for other directories

Reply:
there is a `src/v/storage/fwd.h` but it doesn't have dsk_log_impl 
for me it's ok to use this style of forward declarations, it's a matter of preference I guess